REQUIREMENTS.md

```diff
@@ -0,0 +1,155 @@
+   algorithms, even on potentially untrusted data in a distributed system.
+2. Consistency: Without complex hacks, it's impossible to create link cycles
+   *between* hash-linked blocks. Allowing link cycles one one level (e.g., in
+   relative links within a single block) but not on a larger scale (between
+   blocks) is inconsistent (even though there's nothing we can do about it).
+
+## Stable Pathing
+
+An IPLD path always means the same thing, everywhere, every time. Importantly,
+users shouldn't be able to configure their IPLD library to change how path
+resolution works. They should be able to use alternative path resolution
+algorithms for *non*-IPLD paths (e.g., IPFS paths) but those should be built
+on top of IPLD.
+
+**Motivation:** Deterministic computations on top of a IPLD need to produce the
+same result every time.
+
+## Link Transparent Pathing
+
+Path resolution must transparently traverse links.
+
+**Motivation:** Deduplication and composability. If links are *not* transparent,
+programmers will have to either make many small objects (lots of hashing) or
+inline data into large objects (lots of duplication and copying).
+
+## Multicodecs Are Not Types
+
+It's impossible to understand IPLD data at a *structural* level if we don't know
+the format. Therefore, we should avoid introducing new formats unnecessarily as
+*every* IPLD implementation needs to support these new formats.
+
+## Primitives
+
+The "recommended" IPLD format (currently DagCBOR) needs to support *at a minimum*:
+
+* 32/64 bit integers without losing information.
+* 32/64 bit floats without losing information.
+* Unicode strings.
+* Binary strings.
+* Objects (with string keys, at least).
+* Arrays.
+* Booleans.
+* A bottom type (null).
+
+**Motivation:** Convenience, really.
+
+## Non-Cyclic, Block-Local Relative Links
+
+That is, relative links that don't traverse out the back of an object. See the
+conclusions from: [#1](https://github.com/ipld/specs/issues/1).
+
+**Motivation:** This is required to efficiently represent a highly connected DAG
+of tiny nodes.
+
+**Caveat:** This brings in some sticky issues around mutability. Depending on
+the implementation, relative links within an object may be act like mutable
+links (from the perspective of the user). The concern here is that we don't want
+users to bundle nodes together into single block *because* they want this
+mutability.
+
+# To Do
+
+Working through this, I realized we have a few things we really need to finish a few things before we can
+call IPLD ready.
+
+* **Path links.** Pointers that can only point to objects at block boundaries
+  are useful but severely gimped. We've been fine up till now because we
+  generally don't *edit* complicated datastructures but this will change.
+  ([#83](https://github.com/ipld/specs/issues/83))
+* **Slice links.** For the same reason, we really should support
+  `/ipld/QmID/start..stop` as a syntax for slicing an array. Most programming
```

Comment:
Is this not more of a selectors thing?

I thought links always pointed to one node.

Reply:
It is more of a selector thing. My worry is that most programming languages support this so we should consider it as well.

On the other hand, we could implement this as a higher level type.

Reply:
+1 for higher level type.  That seems to resonate well with the general concept of "advanced layouts" I'm carrying around in my head.  An "advanced layout" object that has some links and offset information in it and acts like an array when viewed through `typed.Node` rather than directly as the Data Model layer data is :ok_hand: